File: UnitOfWork.NET.FS/UnitOfWork.fs
namespace UnitOfWork.NET.FS.Classes

open Autofac
open Caelan.Frameworks.Common.Helpers
open System
open System.Collections
open System.Collections.Generic
open System.Collections.ObjectModel
open System.Linq
open System.Reflection
open UnitOfWork.NET.FS.Interfaces

type UnitOfWork() as uow = 
    let assemblies = ObservableCollection<Assembly>()
    
    let mutable container = 
        let cb = ContainerBuilder()
        cb.Register(fun t -> uow).AsImplementedInterfaces().AsSelf().As<UnitOfWork>() |> ignore
        cb.RegisterType<Repository>().AsSelf().As<IRepository>().PreserveExistingDefaults() |> ignore
        cb.RegisterGeneric(typedefof<Repository<_>>).AsSelf().As(typedefof<IRepository<_>>) |> ignore
        cb.RegisterGeneric(typedefof<Repository<_, _>>).AsSelf().As(typedefof<IRepository<_, _>>) |> ignore
        cb.RegisterGeneric(typedefof<ListRepository<_, _, _>>).AsSelf().As(typedefof<IListRepository<_, _, _>>) |> ignore
        cb.Build()
    
    let isRepository (t : Type) = 
        try 
            t.IsAssignableTo<IRepository>() && t.IsInterface |> not && t.IsAbstract |> not && t <> typeof<Repository> 
            && ((t.IsGenericType && t.GetGenericTypeDefinition() <> typedefof<Repository<_>> && t.GetGenericTypeDefinition() <> typedefof<Repository<_, _>> && t.GetGenericTypeDefinition() <> typedefof<ListRepository<_, _, _>>) || t.IsGenericType |> not) 
            && container.IsRegistered(t) |> not
        with _ -> false
    
    let registerAssembly (assemblyArr : Assembly []) = 
        let cb = ContainerBuilder()
        cb.RegisterAssemblyTypes(assemblyArr |> Array.filter (fun t -> 
                                                    try 
                                                        t.GetTypes() |> Array.exists isRepository
                                                    with _ -> false)).Where(fun t -> t |> isRepository).AsSelf().AsImplementedInterfaces()
        |> ignore
        cb.Update(container)
        assemblyArr
        |> Array.collect (fun t -> t.GetReferencedAssemblies())
        |> Array.map (fun t -> 
               try 
                   t |> Assembly.Load
               with _ -> null)
        |> Array.filter (isNull >> not)
        |> Array.filter (assemblies.Contains >> not)
        |> Array.filter (fun t -> 
               try 
                   t.GetTypes() |> Array.exists isRepository
               with _ -> false)
        |> Array.iter assemblies.Add
    
    do
        let cb = ContainerBuilder()
        let fields = uow.GetType().GetFields().Where(fun t -> t.FieldType |> isRepository) |> Array.ofSeq
        fields |> Array.iter (fun t -> cb.RegisterType(t.FieldType).AsSelf().AsImplementedInterfaces() |> ignore)
        let properties = uow.GetType().GetProperties().Where(fun t -> t.PropertyType |> isRepository) |> Array.ofSeq
        properties |> Array.iter (fun t -> cb.RegisterType(t.PropertyType).AsSelf().AsImplementedInterfaces() |> ignore)
        cb.Update(container)
        fields |> Array.iter (fun t -> t.SetValue(uow, container.ResolveOptional(t.FieldType)))
        properties |> Array.iter (fun t -> t.SetValue(uow, container.ResolveOptional(t.PropertyType)))
        assemblies.CollectionChanged.Add(fun t -> 
            t.NewItems.Cast<Assembly>()
            |> Array.ofSeq
            |> registerAssembly)
        AppDomain.CurrentDomain.GetAssemblies() |> Array.iter assemblies.Add
    
    interface IUnitOfWork with
        member this.Data<'T when 'T : not struct>() = this.Data<'T>()
        member this.CustomRepository<'TRepository when 'TRepository :> IRepository>() = this.CustomRepository<'TRepository>()
        member this.Repository<'T when 'T : not struct>() = this.Repository<'T>()
        member this.Repository<'TSource, 'TDestination when 'TSource : not struct and 'TDestination : not struct>() = this.Repository<'TSource, 'TDestination>()
        member this.Repository<'TSource, 'TDestination, 'TListDestination when 'TSource : not struct and 'TDestination : not struct and 'TListDestination : not struct>() = this.Repository<'TSource, 'TDestination, 'TListDestination>()
    
    interface IDisposable with
        member this.Dispose() = this.Dispose()
    
    member this.RegisterRepository<'TRepository when 'TRepository :> IRepository>() = 
        this.RegisterRepository(typeof<'TRepository>)

    member __.RegisterRepository repositoryType =
        if not <| repositoryType.IsInterface && not <| repositoryType.IsAbstract && not <| container.IsRegistered(repositoryType) then 
            let cb = ContainerBuilder()
            if repositoryType.IsGenericTypeDefinition then
                cb.RegisterGeneric(repositoryType).AsSelf().AsImplementedInterfaces() |> ignore
            else    
                cb.RegisterType(repositoryType).AsSelf().AsImplementedInterfaces() |> ignore
            cb.Update(container)
    
    member private this.GetRepository<'TRepository when 'TRepository :> IRepository>() = 
        this.RegisterRepository<'TRepository>()
        container.Resolve<'TRepository>()
    
    member this.CustomRepository<'TRepository when 'TRepository :> IRepository>() = this.GetRepository<'TRepository>()
    member this.Repository<'T when 'T : not struct>() = this.GetRepository<IRepository<'T>>()
    member this.Repository<'TSource, 'TDestination when 'TSource : not struct and 'TDestination : not struct>() = this.GetRepository<IRepository<'TSource, 'TDestination>>()
    member this.Repository<'TSource, 'TDestination, 'TListDestination when 'TSource : not struct and 'TDestination : not struct and 'TListDestination : not struct>() = this.GetRepository<IListRepository<'TSource, 'TDestination, 'TListDestination>>()
    abstract Data<'T when 'T : not struct> : unit -> seq<'T>
    override __.Data<'T when 'T : not struct>() = Seq.empty<'T>
    
    abstract Dispose : unit -> unit
    override this.Dispose() = ()

File: UnitOfWork.NET/Classes/UnitOfWork.cs
using System;
using System.Reflection;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using Autofac;
using UnitOfWork.NET.Interfaces;
using System.Linq;
using UnitOfWork.NET.Extenders;

namespace UnitOfWork.NET.Classes
{
	public class UnitOfWork : IUnitOfWork
	{
		private readonly ObservableCollection<Assembly> _assemblies;
		private readonly IContainer _container;

		public UnitOfWork()
		{
			_assemblies = new ObservableCollection<Assembly>();

			var cb = new ContainerBuilder();

			cb.Register(t => this).AsImplementedInterfaces().AsSelf().As<UnitOfWork>();
			cb.RegisterType<Repository>().AsSelf().As<IRepository>().PreserveExistingDefaults();
			cb.RegisterGeneric(typeof(Repository<>)).AsSelf().As(typeof(IRepository<>));
			cb.RegisterGeneric(typeof(Repository<,>)).AsSelf().As(typeof(IRepository<,>));
			cb.RegisterGeneric(typeof(Repository<,,>)).AsSelf().As(typeof(IListRepository<,,>));

			_container = cb.Build();

			RegisterProperties();

			_assemblies.CollectionChanged += (sender, args) =>
			{
				RegisterAssembly(args.NewItems.Cast<Assembly>().ToArray());

				UpdateProperties();
			};

			foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies()) _assemblies.Add(assembly);
		}

		protected void RegisterProperties()
		{
			var cb = new ContainerBuilder();

			var fields = GetType().GetFields().ToArray();
			var properties = GetType().GetProperties().ToArray();

			foreach (var type in fields.Select(t => t.FieldType).Union(properties.Select(t => t.PropertyType)).Where(IsRepository)) RegisterRepository(cb, type);

			cb.Update(_container);
		}

		protected void UpdateProperties()
		{
			var fields = GetType().GetFields().ToArray();
			var properties = GetType().GetProperties().ToArray();

			foreach (var field in fields.Where(t => t.FieldType.IsAssignableTo<IRepository>())) field.SetValue(this, _container.ResolveOptional(field.FieldType));
			foreach (var property in properties.Where(t => t.PropertyType.IsAssignableTo<IRepository>())) property.SetValue(this, _container.ResolveOptional(property.PropertyType));
		}

		private bool IsRepository(Type t)
		{
			try
			{
				return t.IsAssignableTo<IRepository>() && !t.IsInterface && !t.IsAbstract && t != typeof(Repository) && ((t.IsGenericType && t.GetGenericTypeDefinition() != typeof(Repository<>) && t.GetGenericTypeDefinition() != typeof(Repository<,>) && t.GetGenericTypeDefinition() != typeof(Repository<,,>)) || !t.IsGenericType) && !_container.IsRegistered(t);
			}
			catch
			{
				return false;
			}
		}

		private void RegisterAssembly(Assembly[] assemblyArr)
		{
			var cb = new ContainerBuilder();

			cb.RegisterAssemblyTypes(assemblyArr.Where(t =>
			{
				try
				{
					return t.GetTypes().Any(IsRepository);
				}
				catch
				{
					return false;
				}
			}).ToArray()).Where(IsRepository).AsSelf().AsImplementedInterfaces();
			cb.Update(_container);

			foreach (var assembly in assemblyArr.SelectMany(t => t.GetReferencedAssemblies()).Select(t =>
			{
				try
				{
					return Assembly.Load(t);
				}
				catch
				{
					return null;
				}
			}).Where(
				t =>
				{
					try
					{
						return t != null && !_assemblies.Contains(t) && t.GetTypes().Any(IsRepository);
					}
					catch
					{
						return false;
					}
				}))
				_assemblies.Add(assembly);
		}

		public void RegisterRepository<TRepository>() where TRepository : IRepository => RegisterRepository(typeof(TRepository));

		public void RegisterRepositories(Type[] repositoryTypes)
		{
			var cb = new ContainerBuilder();

			foreach (var repositoryType in repositoryTypes)
			{
				if (repositoryType.IsInterface || repositoryType.IsAbstract || IsRepositoryRegistered(repositoryType)) return;
				RegisterRepository(cb, repositoryType);
			}

			UpdateContainer(cb);
		}

		protected virtual void RegisterRepository(ContainerBuilder cb, Type repositoryType)
		{
			if (IsRepositoryRegistered(repositoryType)) return;

			if (repositoryType.IsGenericTypeDefinition)
			{
				cb.RegisterGeneric(repositoryType).AsSelf().AsRepository().AsImplementedInterfaces();
			}
			else
			{
				var baseType = repositoryType.BaseType;
				while (baseType.IsAssignableTo<IRepository>() && baseType.GenericTypeArguments.Length <= 0)
				{
					baseType = baseType.BaseType;
				}
				var args = baseType.GenericTypeArguments;
				Type sourceType = null;
				Type destType = null;
				if (args.Length > 0)
				{
					sourceType = args.First();
					if (args.Length > 1)
					{
						destType = args.Skip(1).First();
					}
				}
				cb.RegisterType(repositoryType).AsSelf().AsRepository(sourceType, destType).AsImplementedInterfaces();
			}
		}

		public void RegisterRepository(Type repositoryType)
		{
			RegisterRepositories(new[] { repositoryType });
		}

		private TRepository GetRepository<TRepository>() where TRepository : IRepository
		{
			RegisterRepository<TRepository>();

			return _container.Resolve<TRepository>();
		}

		protected void UpdateContainer(ContainerBuilder cb) => cb.Update(_container);

		protected bool IsRepositoryRegistered<TRepository>() where TRepository : IRepository => IsRepositoryRegistered(typeof(TRepository));

		protected bool IsRepositoryRegistered(Type repositoryType) => _container.IsRegistered(repositoryType);

		public TRepository CustomRepository<TRepository>() where TRepository : IRepository => GetRepository<TRepository>();

		public IRepository<T> Repository<T>() where T : class, new() => GetRepository<IRepository<T>>();

		public IRepository<TSource, TDestination> Repository<TSource, TDestination>() where TSource : class, new() where TDestination : class, new() => GetRepository<IRepository<TSource, TDestination>>();

		public IListRepository<TSource, TDestination, TListDestination> Repository<TSource, TDestination, TListDestination>() where TSource : class, new() where TDestination : class, new() where TListDestination : class, new() => GetRepository<IListRepository<TSource, TDestination, TListDestination>>();

		public virtual IQueryable<T> Data<T>() where T : class, new() => Enumerable.Empty<T>().AsQueryable();

		public virtual void Dispose()
		{
			_container.Dispose();
		}
	}
}

